src/main/java/games/strategy/persistence/serializable/DefaultProxyFactoryRegistry.java

```diff
@@ -0,0 +1,44 @@
+package games.strategy.persistence.serializable;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+final class DefaultProxyFactoryRegistry implements ProxyFactoryRegistry {
+  private final Map<Class<?>, ProxyFactory> proxyFactoriesByPrincipalType = new HashMap<>();
+
+  @Override
+  public Optional<ProxyFactory> getProxyFactory(final Class<?> principalType) {
```

Comment:
It may not be clear from the client (`g.s.persistence.serializable.ObjectOutputStream`), so let me elaborate a bit.

We query the `ProxyFactoryRegistry` for _every_ class that passes through the stream.  Those classes include not only the domain-specific classes for which we have provided proxies, but also the JDK classes (among others) that those domain-specific classes are built upon.  There will be no proxy factory registered for those classes as they tend to be `Serializable`.  The presence of a proxy factory simply gives us the ability to override the default serialization behavior.  If one doesn't exist, it is not necessarily an error, as we simply pass the object along and let the JDK Serialization framework handle it as it normally would.

Irrespective of that, if the goal is to get rid of the `Optional` return type, I see two possibilities.

First is what you suggested above.  Adding a `hasProxyFactory()` method and throwing an exception from `getProxyFactory()` if one isn't registered for the specified class.  That would lead to client code in `ObjectOutputStream` like:

```java
@Override
protected Object replaceObject(final Object obj) {
  if (obj == null) {
    return null;
  }

  return proxyFactoryRegistry.hasProxyFactory(obj.getClass())
      ? proxyFactoryRegistry.getProxyFactory(obj.getClass()).newProxyFor(obj)
      : obj;
}
```

The second possibility is to return a Null Object from `getProxyFactory()` in the event a proxy factory has not been registered.  The Null Object in this case would simply be the "identity" factory that returns its argument:

```java
@Override
public ProxyFactory getProxyFactory(final Class<?> principalType) {
  checkNotNull(principalType);

  return proxyFactoriesByPrincipalType.containsKey(principalType)
      ? proxyFactoriesByPrincipalType.get(principalType)
      : ProxyFactory.newInstance(principalType, Function.identity());
}
```

Then the `ObjectOutputStream` code simplifies to

```java
@Override
protected Object replaceObject(final Object obj) {
  return (obj != null) ? proxyFactoryRegistry.getProxyFactory(obj.getClass()) : null;
}
```